### PR TITLE
Calculcation of MAX_THREADS for Parallel Runner has been customized

### PR DESCRIPTION
**Pull request by toktarev**
Calculcation of MAX_THREADS for Parallel Runner has been customized.

### CODE REVIEW

File: hazelcast/src/test/java/com/hazelcast/test/annotation/TestProperties.java [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+
+import java.lang.annotation.Target;
+import java.lang.annotation.Retention;
+import java.lang.annotation.ElementType;
+import java.lang.annotation.RetentionPolicy;
+
+@Target({ElementType.TYPE})
+@Retention(RetentionPolicy.RUNTIME)
+public @interface TestProperties {
+    String maxThreadsCalculatorClass();
```
Comment (mdogan) [inline hazelcast/src/test/java/com/hazelcast/test/annotation/TestProperties.java]: Why not `Class<? extends MaxThreadsAware> maxThreadsCalculatorClass();`?

Comment (mdogan) [inline hazelcast/src/test/java/com/hazelcast/test/annotation/TestProperties.java]: That way `HazelcastParallelClassRunner.getMaxThreads()` method will be simply:

```java
private int getMaxThreads(Class<?> klass) {
    TestProperties properties = klass.getAnnotation(TestProperties.class);

    if (properties != null) {
        Class<? extends MaxThreadsAware> clazz = properties.maxThreadsCalculatorClass();
        try {
            Constructor c = clazz.getConstructor();
            MaxThreadsAware maxThreadsAware = (MaxThreadsAware) c.newInstance();
            return maxThreadsAware.maxThreads();
        } catch (Throwable e) {
            return MAX_THREADS;
        }
    } else {
        return MAX_THREADS;
    }
}
```

File: hazelcast/src/test/java/com/hazelcast/test/MaxThreadsAware.java [hunk 2]
```diff
@@ -0,0 +1,5 @@
+package com.hazelcast.test;
+
+public interface MaxThreadsAware {
```
Comment (mdogan) [inline hazelcast/src/test/java/com/hazelcast/test/MaxThreadsAware.java]: Can you add a javadoc to explain purpose and usage of this interface?

File: hazelcast/src/test/java/com/hazelcast/test/annotation/TestProperties.java (hunk 1 above)
Comment (toktarev) [inline hazelcast/src/test/java/com/hazelcast/test/annotation/TestProperties.java]: In this case you will have to separatelly implement maxThreads callucaltion for each new test.
In this case you will be able to refer on existent calculator

File: hazelcast/src/test/java/com/hazelcast/test/MaxThreadsAware.java (hunk 2 above)
Comment (toktarev) [inline hazelcast/src/test/java/com/hazelcast/test/MaxThreadsAware.java]: Sure second

File: hazelcast/src/test/java/com/hazelcast/test/annotation/TestProperties.java (hunk 1 above)
Comment (mdogan) [inline hazelcast/src/test/java/com/hazelcast/test/annotation/TestProperties.java]: No, maxThreads will be the same. What I suggest is to change signature of annotation to:

```java

@Target({ElementType.TYPE})
@Retention(RetentionPolicy.RUNTIME)
public @interface TestProperties {
    Class<? extends MaxThreadsAware> maxThreadsCalculatorClass();
}
```
That way, you don't need to do `Class.forName()` from class-name and cast it etc..

Comment (toktarev) [inline hazelcast/src/test/java/com/hazelcast/test/annotation/TestProperties.java]: I see good idea

Comment (toktarev) [inline hazelcast/src/test/java/com/hazelcast/test/annotation/TestProperties.java]: Fixed

File: hazelcast/src/test/java/com/hazelcast/mapreduce/MapReduceTestThreads.java [hunk 3]
```diff
@@ -0,0 +1,11 @@
+package com.hazelcast.mapreduce;
+
+import com.hazelcast.test.MaxThreadsAware;
+import com.hazelcast.test.TestEnvironment;
+
+public class MapReduceTestThreads implements MaxThreadsAware {
+    @Override
+    public int maxThreads() {
+        return !TestEnvironment.isMockNetwork() ? 1 : Math.max(Runtime.getRuntime().availableProcessors() / 2, 1);
```
Comment (jerrinot) [inline hazelcast/src/test/java/com/hazelcast/mapreduce/MapReduceTestThreads.java]: Implementations of `MaxThreadsAware` should not deal with the `TestEnvironment.isMockNetwork()`. It should remain a responsibility of the ParallelRunner. 

When the MockNetwork is disabled then parallel runner should always use 1 thread.

Comment (toktarev) [inline hazelcast/src/test/java/com/hazelcast/mapreduce/MapReduceTestThreads.java]: Fixed

File: hazelcast/src/test/java/com/hazelcast/test/MaxThreadsAware.java [hunk 4]
```diff
@@ -0,0 +1,14 @@
+package com.hazelcast.test;
+
+/**
+ *  Interface to implement maximum threads calculation for each thread separately
```
Comment (jerrinot) [inline hazelcast/src/test/java/com/hazelcast/test/MaxThreadsAware.java]: typo? "for each *test* separately"

Comment (toktarev) [inline hazelcast/src/test/java/com/hazelcast/test/MaxThreadsAware.java]: yep

File: hazelcast/src/test/java/com/hazelcast/test/MaxThreadsAware.java [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+ *  Use
+ *
+ *  @TestProperties(maxThreadsCalculatorClass="MYouraxThreadsAwareImpl")
```
Comment (jerrinot) [inline hazelcast/src/test/java/com/hazelcast/test/MaxThreadsAware.java]: typo? `MYouraxThreadsAwareImpl`? what's that?

File: hazelcast/src/test/java/com/hazelcast/test/HazelcastParallelClassRunner.java [hunk 6]
```diff
@@ -28,23 +30,46 @@
...
     }
 
     public HazelcastParallelClassRunner(Class<?> klass, Object[] parameters,
                                         String name) throws InitializationError {
         super(klass, parameters, name);
+        maxThreads =  getMaxThreads(klass);
+    }
+
+    private int getMaxThreads(Class<?> klass) {
+        if (!TestEnvironment.isMockNetwork())
```
Comment (jerrinot) [inline hazelcast/src/test/java/com/hazelcast/test/HazelcastParallelClassRunner.java]: please always use the curly brackets for branches.

**Comment by jerrinot** [comment]
:+1: once it's green

**Comment by mdogan** [comment]
:+1:

**Pull request merged**